### ISSUE TITLE
[GitHub] Current state of PRs

### ISSUE THREAD

**Issue by dimakuv** [oscarlab/graphene#658]
This gives an overview of the current state of PRs. Note that I use terms "exception", "interrupt", "signal" interchangeably.

Legend:
- "not reviewed": needs more reviews!
- "not finished": needs actions from the author, not updated recently
- "WIP": actively updated and reviewed
- "depends on #PR": other PRs need to merged first
- "conditioned on #PR": if other PRs are merged, this PR must be abandoned

# Large non-critical PRs

- #229 by @chiache, not finished (*performance*: improving handle polling semantics)
- #234 by @rainfld, not finished/reviewed (*new feature*: dynamic memory management features (EDMM) with SGX2)
- #291 by @yamahata, not finished/reviewed (*performance*: fork optimization by cross memory attach)
- #332 by @chiache, not finished (*security*: local attestation protocol for process creation on SGX)
- #405 by @dimakuv, not reviewed (*performance*: exitless system calls)

# Small PRs

- #197 by @donporter, not reviewed (socket handling waiting fixes)
    - needs patch #434 by @thomasknauth, not reviewed
    - *ignore* if #229 by @chiache is merged
- #527 by @chiache, WIP (ensure the usage of alternative stack for host exception handling)
- #556 by @yamahata, WIP (preparation to use %gs for libos tcb)
- #591 by @zhangckid, WIP (optimize do_exit_group for multi thread race condition)
- #601 by @yamahata, WIP (make shim_tcb_t in glibc tcb compile time option)
     - depends on #555 and #556
- #620 by @HW42 (add memory gap around 'exec' area)

# Debates

(We need to decide on the general direction with these PRs.)

## Floating Point (FP) support
- #301 by @yamahata (Graphene doesn't save/restore FP context on emulated syscalls)
    - PR forces the compiler to generate no-FP assembly in LibOS/PAL layers (except crypto lib)
    - Alternative (proposed by @dimakuv): save FP context before syscall emulation and restore it afterwards; measure performance and then decide
- #397 by @yamahata (Graphene doesn't propagate FP context through PAL layer on exceptions)

## Multi-threaded issues and races:
- #322 by @yamahata, WIP (race condition on thread exiting at LibOS layer)
- #452 by @yamahata, *to close???* (test program for sigsuspend and alarm)
- #591 by @zhangckid, WIP (optimize do_exit_group for multi thread race condition)

## Signal handling:
- Info: https://github.com/oscarlab/graphene/wiki/Signal-Handling-in-Graphene
- Fixes for current implementation:
    - #406 by @yamahata, WIP (disallow nested asynchronous signals by blocking signals instead of queuing them in Linux PAL)
- New proposal:
    - #347 by @yamahata (new signal-handling mechanism at LibOS layer)
    - #632 by @yamahata (bugfixes and dedicated stack at Linux-SGX PAL)

**Comment by dimakuv** [comment]
Interesting dependency chains (from first-to-merge to last-to-merge):

* #555 -> #556 -> #601 (re-using glibc's TCB in LibOS/PAL)

* #619 -> #620 (exec area in the middle of enclave heap, memory gaps)

* #492 -> #475 (testing unmapped memory under Linux-SGX)

* #523 -> #450 and #455 (`SIGNAL_DELAYED` and data races on signal handling)

* #655 -> removes #652 and changes #653 (16B alignment of stack)

* #434 -> #197 or #229 (supersedes both)

**Comment by HW42** [comment]
@dimakuv:
> - #639 by @yamahata, not reviewed (disallow nested async host exception in enclave_entry)
>      - @HW42 @donporter: *please review*

I already commented there a while back. Do you want a more explicit approve comment or do you mean another PR.

**Comment by dimakuv** [comment]
@HW42, I always forget you cannot approve PRs. Sorry, but yes, I wanted an explicit approval. LGTM is good enough.

**Comment by dimakuv** [comment]
All these issues were fixed or are active PRs. (All except EDMM/SGX2.) Closing this issue.
